src/mach.js

```diff
@@ -71,21 +212,40 @@ const mach = {
   machCanvasSetSize(canvas, width, height) {
     const cv = mach.canvases[canvas];
     if (width > 0 && height > 0) {
-      cv.canvas.width = width;
-      cv.canvas.height = height;
+      cv.canvas.style.width = width + "px";
+      cv.canvas.style.height = height + "px";
+      cv.canvas.width = width * window.devicePixelRatio;
+      cv.canvas.height = height * window.devicePixelRatio;
     }
   },
 
-  machCanvasGetWidth(canvas) {
+  machCanvasGetWindowWidth(canvas) {
+    const cv = mach.canvases[canvas];
+    return cv.canvas.width / window.devicePixelRatio;
+  },
+
+  machCanvasGetWindowHeight(canvas) {
+    const cv = mach.canvases[canvas];
+    return cv.canvas.height / window.devicePixelRatio;
+  },
+
+  machCanvasGetFramebufferWidth(canvas) {
     const cv = mach.canvases[canvas];
     return cv.canvas.width;
   },
 
-  machCanvasGetHeight(canvas) {
+  machCanvasGetFramebufferHeight(canvas) {
     const cv = mach.canvases[canvas];
     return cv.canvas.height;
   },
 
+  machEventShift() {
+    if (mach.events.length < 0)
+      return 0;
+
+    return mach.events.shift();
```

Comment:
Note to self: Possible optimization: Use an array and a cursor indicating the current head of the list. Each time we call this function, return list[cursor] and then increment it. Finally reset the list and cursor at the end of frame (needs a new function)

The point is that shift operation may involve moving of all array elements which is performance intensive. In practice, it is likely that JS engines optimize this properly.